### PR TITLE
feat(todomvc): refactor to not expose too much API and add local storage support

### PR DESCRIPTION
**Pull request by christianalfoni**
- I have just simplified the use of the API to not introduce too many new concepts
- I have inlined all signals in controller to better give an impression of how you can describe the app logic with operators
- I have put controller code with comments into docs
- I have added a new local storage provider
- Written docs for the provider as well

### CODE REVIEW

File: packages/cerebral-provider-localstorage/README.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# cerebral-provider-http
```
Comment (henri-hulski) [inline packages/cerebral-provider-localstorage/README.md]: Should be cerebral-provider-localstorage.

**Comment by Guria** [comment]
@christianalfoni you did it again. :) Why not split into atomic commits? 1. introduce localstorage provider 2. refactor todomvc

**Comment by Guria** [comment]
another question is why only localStorage? what about sessionStorage? I suppose that creating cerebral-provider-sessionstorage would be overkill, since it has exactly same api

**Comment by Guria** [comment]
btw, manual storage sync looks ugly. haven't you tried to make module instead that would track defined paths in config and will sync it on mutations automatically?

**Comment by christianalfoni** [comment]
@Guria Haha, sorry. Like... I start refactoring and then I see... oh, we should use local storage instead... then I have made so many changes. Stashing those, implement provider and test it, just to pop the stash and figure out that it does not work, stash back, rebase provider commit to fix it... and then all again possibily, it is just such a hassle for me. But yeah... I do not know how to make me better at structuring this process :) I need help! haha

But to other comments:

`cerebral-provider-storage` makes a lot more sense. I can implement localStorage and session for now :)

Also:

```js
StorageProvider({
  sync: {
    todos: 'todos'
  }
})
```

Is a great idea, listen to flushes and automate it :D

Let me make some adjustments here

**Comment by Guria** [comment]
@christianalfoni just don't put all files in one commit when you are in such situations. You can `git add` exact package and related changes (like lerna, Travis setup), commit, and then commit a rest. Or continue to split in the same way.

**Comment by christianalfoni** [comment]
You know, that sounds like something I could do. I get my freedom to make changes by heart and rather when I prepare for commiting I rather pick and choose. These changes are related to that, these over here to that other thing. And if some of them are mixed I can just commit them at the end? :)

**Comment by bfitch** [comment]
@christianalfoni since you know I love laziness...

What I often do when I'm ready open a PR, is I `git reset` to the commit SHA I started on when first checking out the branch. Then, I group my work into logical commits:

```bash
git add providers/localStorageProvider.js tests/localStorageProviderTest.js
git commit -m "Add localStorage provider."

git add lerna.js travis.yml`
git commit -m "Update CI and deploy config"
```
Now you have one isolated commit for all the localStorage stuff and one for the deploy stuff. Now if @Guria tells you to pull out docs changes or CI or whatever, it should be easy to `git cherry-pick` onto a new branch.

With this flow, you don't have to think about your commits or grouping things for git **while coding**. You can keep hacking without interruption or stashing, but you can clean everything up with minimum fuss and easily revert things.

**Comment by christianalfoni** [comment]
Okay, so I refactored to sync and sessionStorage added. That sync thing is awesome :)

@henri-hulski I kept `uid`, cause it really means "client side id". Its exact meaning I am not sure of. Though `id` is typically related to server side, and uid is something client creates. But yeah, experience and opinions :) No worries if you want to PR and change it, happy to approve... I just have a glass of wine too much in me right now 🍷 :)

File: packages/cerebral-provider-storage/src/index.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
+function LocalStorageProvider (options = {}) {
```
Comment (henri-hulski) [inline packages/cerebral-provider-storage/src/index.js]: LocalStorageProvider -> StorageProvider ?

File: packages/cerebral-provider-storage/src/index.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    if (context.debugger) {
+      context.debugger.wrapProvider('storage')
+    }
+
+    return context
+  }
+}
+
+export default LocalStorageProvider
```
Comment (henri-hulski) [inline packages/cerebral-provider-storage/src/index.js]: Same here

File: packages/cerebral-provider-storage/README.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# cerebral-provider-http
+
+Go to [Cerebral official site](https://cerebral.github.io/docs/packages/localstorage.html) to read docs.
```
Comment (henri-hulski) [inline packages/cerebral-provider-storage/README.md]: localstorage.html -> storage.html

File: packages/cerebral-provider-storage/README.md [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# cerebral-provider-http
```
Comment (henri-hulski) [inline packages/cerebral-provider-storage/README.md]: cerebral-provider-http -> cerebral-provider-storage

**Comment by henri-hulski** [comment]
Took another look at sync.
Quite simple implementation.
I really like it!

File: docs/docs/docs/get_started/todomvc.md [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    ],
+    newTodoTitleChanged: [
+      set(state`newTodoTitle`, props`title`)
+    ],
+    newTodoSubmitted: [
+      when(state`newTodoTitle`), {
+        true: [
+          addTodo,
+          set(state`newTodoTitle`, ''),
+          setLocalStorage('todos', state`todos`)
```
Comment (Guria) [inline docs/docs/docs/get_started/todomvc.md]: suppose this changed now

File: docs/docs/docs/packages/storage.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+### instantiate
+
+```js
+import {Controller} from 'cerebral'
+import StorageProvider from 'cerebral-provider-storage'
+
+const controller = Controller({
+  providers: [StorageProvider({
+    target: 'localStorage' // default, "sessionStorage" is alternative
```
Comment (Guria) [inline docs/docs/docs/packages/storage.md]: I'd like to have `prefix` option to prepend to every storage key

**Pull request merged**